src/uct/tcp/tcp_ep.c

```diff
@@ -1038,6 +1269,64 @@ ucs_status_t uct_tcp_ep_am_zcopy(uct_ep_h uct_ep, uint8_t am_id, const void *hea
     return status;
 }
 
+ucs_status_t uct_tcp_ep_put_zcopy(uct_ep_h uct_ep, const uct_iov_t *iov,
+                                  size_t iovcnt, uint64_t remote_addr,
+                                  uct_rkey_t rkey, uct_completion_t *comp)
+{
+    uct_tcp_ep_t *ep             = ucs_derived_of(uct_ep, uct_tcp_ep_t);
+    uct_tcp_iface_t *iface       = ucs_derived_of(uct_ep->iface, uct_tcp_iface_t);
+    uct_tcp_am_hdr_t *hdr        = NULL;
+    uct_tcp_ep_put_req_t put_req = { 0 };
+    uct_tcp_ep_zcopy_tx_t *ctx;
+    ucs_status_t status;
+
+    status = uct_tcp_ep_prepare_zcopy(iface, ep, 0, &put_req, sizeof(put_req),
+                                      iov, iovcnt, "put_zcopy",
+                                      UCT_TCP_EP_PUT_ZCOPY_MAX -
+                                      sizeof(uct_tcp_am_hdr_t),
+                                      /* Set a payload length directly to the
+                                       * TX length, since PUT Zcopy doesn't
+                                       * set the payload length to TCP AM hdr */
+                                      &ep->tx.length, &hdr);
+    if (ucs_unlikely(status != UCS_OK)) {
+        return status;
+    }
+
+    ctx                 = ucs_derived_of(hdr, uct_tcp_ep_zcopy_tx_t);
+    hdr->am_id          = UCT_TCP_EP_PUT_REQ_AM_ID;
+    hdr->length         = sizeof(put_req);
+    put_req.addr        = remote_addr;
+    put_req.length      = ep->tx.length;
+    put_req.remote_comp = (uint64_t)(uintptr_t)comp;
+
+    status = uct_tcp_ep_am_sendv(iface, ep, 0, hdr, UCT_TCP_EP_PUT_ZCOPY_MAX,
+                                 &put_req, ctx->iov, ctx->iov_cnt);
+    if (ucs_likely((status == UCS_OK) || (status == UCS_ERR_NO_PROGRESS))) {
+        UCT_TL_EP_STAT_OP(&ep->super, PUT, ZCOPY, put_req.length);
+
+        if (uct_tcp_ep_ctx_buf_need_progress(&ep->tx)) {
+            /* Since EP has to wait for an acknowledgment for this PUT
```

Comment:
why do we need to wait for acknowledgement?

Reply:
there are gtests that check a remote data upon completion of a PUT operation, i.e. thay assume that a remote buffer has to contain a PUT data

Reply:
hmm, flush should be used for ensuring rma remote completion. 
@yosefe, wdyt?

Reply:
@brminich sorry, you're right, this checked after flush(). But if we don't have an ack for the PUT operation, we don't know when we should return OK from flush() - i.e. decrement `iface:: outstanding` upon PUT ack receiving from a peer. TCP completes a UCT completion and makes flush() returning OK at the same time

Reply:
Then, I'd do it as follows:
- Use real completion for UCT put
- Maintain a counter of unacknowledged put operations
- Check this counter in flush operations

Reply:
@brminich nice catch
TCP does this sequence, but EP flush doesn't work correctly (it returns `UCS_OK` when iface return `UCS_ERR_NO_PROGRESS` due to a PUT operation on an EP).